### PR TITLE
Add serde bounds to zebra-chain structures.

### LINKED ISSUE #137: Make sure all of our types are Serde'able.
So that they can be serialized/deserialized via Serde to our storage.

### LINKED ISSUE #195: Add `Serialize`, `Deserialize` bounds to `Transaction` and other structs.
Adding `Serialize` and `Deserialize` bounds to `Transaction` and other structs (`Block`?) in `zebra-chain` now will ensure that all of their component types are serde'able as we continue fleshing them out.

### PR DESCRIPTION
**Pull request by hdevalence**
Closes #195

### CODE REVIEW

**Comment by hdevalence** [comment]
Blocked on #249 since we need to add serde bounds to ciphertext types.

File: zebra-chain/src/transaction/shielded_data.rs
Comment (hdevalence) [inline zebra-chain/src/transaction/shielded_data.rs]: Could this be called something that indicates it's a Serde helper, like `AffinePointHelper`?

File: zebra-chain/src/transaction/shielded_data.rs
Comment (hdevalence) [inline zebra-chain/src/transaction/shielded_data.rs]: `LocalEither` seems fine but it would be good to be consistent with whatever naming scheme was used above (previous comment)

File: zebra-chain/src/lib.rs
Comment (hdevalence) [inline zebra-chain/src/lib.rs]: I'm fine with using `serde_big_array` if other people are, but I think we should use it consistently if we use it at all, there are manual Deserialize implementations for BCTV14 and Groth16 proofs that should be removed if we opt to use it.

**Comment by hdevalence** [comment]
LGTM modulo comments (but I can't be the reviewer since I opened the PR)

File: zebra-chain/src/transaction/shielded_data.rs
Comment (dconnolly) [reply zebra-chain/src/transaction/shielded_data.rs]: Also `jubjub::AffinePoint` is used in several `sapling` key types, if those ever get serialized (for wallet-related stuff) having the same serde helper used for all instances would be good.

**Review by dconnolly — DISMISSED** [review]
🐌

File: zebra-chain/src/lib.rs
Comment (teor2345) [inline zebra-chain/src/lib.rs]: Should we use the const from equihash_solution instead?

File: zebra-chain/src/lib.rs
Comment (teor2345) [inline zebra-chain/src/lib.rs]: It might be helpful to explain which structs need arrays of size 580, 601, and 80.

File: zebra-chain/src/transaction/shielded_data.rs
Comment (yaahc) [reply zebra-chain/src/transaction/shielded_data.rs]: How about I make a `serde_helpers` module and move all of these types into that centralized location.

**Comment by yaahc** [comment]
I want to see what automation does but when I run the build locally from `zebra-chain` it fails but when I run it from the root directory of the entire project `zebra-chain` compiles successfully along with the rest of the project. I don't necessarily think this is a problem, its probably just different features being enabled by default vs by zebrad, but if this compile error isn't caught by automation then we will want to fix that as well.

File: zebra-chain/src/lib.rs
Comment (yaahc) [reply zebra-chain/src/lib.rs]: Oh, do you mean move it back to the `equihash_solution` module and make it pub(crate) and use it's full path in `serde_helpers` when referencing it?

File: zebra-chain/src/lib.rs
Comment (dconnolly) [inline zebra-chain/src/lib.rs:10]: 👍

File: zebra-chain/src/transaction/shielded_data.rs
Comment (dconnolly) [inline zebra-chain/src/transaction/shielded_data.rs:83]: 👍

**Review by dconnolly — DISMISSED** [review]
👏

**Review by yaahc — DISMISSED** [review]
I think this is good to go

**Review by yaahc — APPROVED** [review]
reapprove after fixing lockfile conflict

**Pull request merged**
